core/src/main/java/de/javakaffee/web/msm/storage/RedisStorageClient.java

```diff
@@ -0,0 +1,269 @@
+ * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package de.javakaffee.web.msm.storage;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+import org.apache.juli.logging.Log;
+import org.apache.juli.logging.LogFactory;
+
+import redis.clients.jedis.BinaryJedis;
+import redis.clients.jedis.exceptions.JedisConnectionException;
+
+/**
+ * <p>
+ * Storage client backed by a Jedis client instance.
+ * </p>
+ * <p>
+ * Note that although all functions return {@link Future} instances for compatibility with the memcached client,
+ * all Redis commands are currently actually executed synchronously.
+ * </p>
+ */
+public class RedisStorageClient implements StorageClient {
+    protected static final Log _log = LogFactory.getLog(RedisStorageClient.class);
+
+    private final String _host;
+    private final int _port;
+    private final boolean _ssl;
+    private final JedisPool _pool = new JedisPool();
+    
+    /**
+     * Creates a <code>MemcachedStorageClient</code> instance which connects to the given Redis URL.
+     * 
+     * @param redisUrl redis URL 
+     */
+    public RedisStorageClient(String redisUrl) {
+        if (redisUrl == null)
+            throw new NullPointerException("Param \"redisUrl\" may not be null");
+        
+        if (_log.isDebugEnabled())
+            _log.debug(String.format("Creating RedisStorageClient with URL \"%s\"", redisUrl));
+
+        // Support a redis URL in the form "redis://hostname:port" like the client "Lettuce" does
+        if (!(redisUrl.startsWith("redis://") || redisUrl.startsWith("rediss://")))
+            throw new IllegalArgumentException("Redis URL must start with \"redis://\" or \"rediss://\"");
+        
+        _ssl = redisUrl.startsWith("rediss://");
+
+        String hostNamePort = redisUrl.substring(redisUrl.indexOf('/') + 2);
+        int pos = hostNamePort.indexOf(':');
+        if (pos != -1) {
+            _host = hostNamePort.substring(0, pos);
+            try {
+                _port = Integer.parseInt(hostNamePort.substring(pos + 1));
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Error parsing port number in Redis URL");
+            }
+        } else {
+            _host = hostNamePort;
+            _port = 6379;
+        }
+    }
+    
+    @Override
+    public boolean add(final String key, final int exp, final byte[] o) throws IOException {
+        
+        if (_log.isDebugEnabled())
+            _log.debug(String.format("Adding key to Redis (key=%s, exp=%d, %d bytes)", key, exp, o.length));
+
+        return (new JedisCommandRunner<Boolean>() {
+            private boolean _setCompleted;
```

Comment:
Shouldn't this be volatile? Or is it guaranteed that this is run again by the same thread if connection failed?